File: pages/refactorings/replace-global-state-with-parameter/03-fresh-names/After.scala
// Number every variable in an expression tree, pre-order, with a
// fresh-number supply.
object After:
  enum Expr:
    case Var(name: String)
    case Lit(value: Int)
    case Add(lhs: Expr, rhs: Expr)

  def number(e: Expr): Expr = go(e, 0)._1

  private def go(e: Expr, n: Int): (Expr, Int) = e match
    case Expr.Var(name) => (Expr.Var(name + n), n + 1)
    case Expr.Lit(_)    => (e, n)
    case Expr.Add(l, r) =>
      val (l1, n1) = go(l, n)
      val (r1, n2) = go(r, n1)
      (Expr.Add(l1, r1), n2)

File: pages/refactorings/replace-global-state-with-parameter/03-fresh-names/After.hs
-- Number every variable in an expression tree, pre-order, with a
-- fresh-number supply.
module After where

data Expr = Var String | Lit Int | Add Expr Expr
  deriving (Eq, Show)

number :: Expr -> Expr
number e = fst (go e 0)

go :: Expr -> Int -> (Expr, Int)
go (Var name) n = (Var (name ++ show n), n + 1)
go e@(Lit _) n = (e, n)
go (Add l r) n =
  let (l', n') = go l n
      (r', n'') = go r n'
  in (Add l' r', n'')
